EIPS/eip-6963.md

```diff
@@ -0,0 +1,118 @@
+---
+eip: 6963
+title: Multi Injected Provider Interface (aka MIPI)
+description: Using `window.evmproviders` instead of `window.ethereum`
+author: Pedro Gomes (@pedrouid), Kosala Hemachandra (@kvhnuke), Richard Moore (@ricmoo), Gregory Markou (@GregTheGreek)
+discussions-to: https://ethereum-magicians.org/t/eip-6963-multi-injected-provider-interface-aka-mipi/14076/2
+status: Draft
+type: Standards Track
+category: Interface
+created: 2023-05-01
+requires: 1193
+---
+
+## Abstract
+
+A Javascript interface injected by browser extensions installed by the user that enables user selection from multiple wallet providers using a window object labelled `window.evemproviders` instead of overwriting `window.ethereum` object.
+
+## Motivation
+
+Currently, wallet providers that offer browser extensions must inject their Ethereum providers (EIP-1193) into the same window object `window.ethereum` however this creates conflicts for users that may install more than one browser extension.
+
+Browser extensions are loaded in the web page and do not have a predictable pattern resulting in a race condition where the user is not in control to choose which wallet provider must be selected for exposing Ehtereum interface under the `window.ethereum` object.
+
+This results not only in a degraded user experience but also increases the barrier to entry for new browser extensions as users are forced to only install one browser extension at a time.
+
+Some browser extensions attempt to counteract this problem by delaying their injection to overwrite the same `window.ethereum` object which creates an unfair competition for wallet providers and lack of interoperability.
+
+In this proposal, we provide a solution that optimitizes for interoperability and enables fairer competition by lowering the barrier to entry for new wallet providers to improve the user experience on Ethereum.
+
+This is achieved by replacing the current window object `window.ethereum` with a new window object `window.evmproviders` that enables multiple browser extensions to be exposed for user selection in the web page.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Provider Info
+
+Each wallet provider will be exposed with the following interface `ProviderInfo` that will be used to display to the user:
+
+```typescript
+/**
+ * Represents the assets needed to display a wallet
+ */
+interface ProviderInfo {
+  uuid: string;
+  walletId: string;
```

Comment:
Given the widespread misuse of `window.ethereum.isMetaMask = true`, are there any potential mitigations for wallets that just use the MetaMask `walletId` in their `ProviderInfo` and hope to inject first to be picked up by the wallet switcher?

It's not novel to this proposal - EIP-5749 would have been _more_ sensitive to this, but given that it's an array, the wallet switcher will somehow need to choose (for example, when the user clicks a "Connect with MetaMask" button in the UI  and `window.evmproviders.find(provider => provider.info.walletId === 'metamask')` feels like a default solution in the absence of mitigation.

There's a lot of incentive for a wallet provider to continue to duck-type as MetaMask in order to be prioritized (vs buried deep in a wallet selection screen) and with this EIP formalizing the notion of wallet providers whereas the previous standard did not, I'd be curious to mitigate in some way in a way that wasn't necessary previously.

Reply:
I'm not sure if I understand, do you say wallets would still be trying to impersonate MetaMask?

I think unless you're trying to onboard users to install a specific wallet you can simply map through the whole array and display what the user has installed, there's no need to know what wallets your Dapp is displaying, as long as you can get a provider following 1193

Maybe this issue would depend on how Dapps would implement this

Reply:
Actually I was thinking we could do differently and just use reversed domains here instead

com.metamask
com.coinbase
io.rabby
xyz.taho
me.rainbow
com.spot-wallet